unity/unity-embed-host/CoreCLRHost.cs

```diff
@@ -41,15 +52,139 @@ internal static void InitState()
     public static IntPtr /*Assembly*/ load_assembly_from_data(byte* data, long size)
     {
         var assembly = alcWrapper.CallLoadFromAssemblyData(data, size);
-        return (IntPtr)assemblyHandleField.GetValue(assembly);
+        var ptr = GetHandleForAssembly(assembly);
+        return ptr;
     }
 
     [NativeFunction(NativeFunctionOptions.DoNotGenerate)]
     public static IntPtr /*Assembly*/ load_assembly_from_path(byte* path, int length)
     {
         var assembly = alcWrapper.CallLoadFromAssemblyPath(Encoding.UTF8.GetString(path, length));
-        return (IntPtr)assemblyHandleField.GetValue(assembly);
+        var ptr = GetHandleForAssembly(assembly);
+        return ptr;
+    }
+
+    private static IntPtr GetHandleForAssembly(Assembly assembly)
+    {
+        if (!m_assemblies.TryGetValue(assembly, out var pair))
+        {
+            pair = new AssemblyPair() { handle = GCHandle.ToIntPtr(GCHandle.Alloc(assembly)), name = assembly.GetName().Name};
+            m_assemblies.Add(assembly, pair);
+        }
+
+        return pair.handle;
+    }
+
+    private static Assembly GetAssemblyForHandle(IntPtr ptr)
+    {
+        return (Assembly)GCHandle.FromIntPtr(ptr).Target;
+    }
+
+    [NativeFunction(NativeFunctionOptions.DoNotGenerate)]
+    [return: NativeCallbackType("MonoClass*")]
+    public static IntPtr class_from_name(
+        [NativeCallbackType("MonoImage*")] IntPtr image,
+        [NativeCallbackType("const char*")] sbyte* name_space,
+        [NativeCallbackType("const char*")] sbyte* name,
+        bool ignoreCase)
+    {
+        Assembly assembly = GetAssemblyForHandle(image);
+        var ns = new string(name_space);
+        var klass_name = new string(name);
+        string assemblyQualified = $"{ns}.{klass_name}";
+        assemblyQualified = assemblyQualified.Replace('/', '+');
+
+        return assembly.GetType(assemblyQualified, false, ignoreCase).TypeHandleIntPtr();
+    }
+
+    public static int image_get_table_rows([NativeCallbackType("MonoImage*")] IntPtr image, int table_id)
+    {
+        const int MONO_TABLE_TYPEDEF = 2;
+        if (table_id == MONO_TABLE_TYPEDEF)
+        {
+            Assembly assembly = GetAssemblyForHandle(image);
+            return assembly.GetTypes().Length;
+        }
+
+        return 0;
+    }
+
+    [return: NativeCallbackType("MonoClass*")]
+    public static IntPtr unity_class_get([NativeCallbackType("MonoImage*")] IntPtr image, uint type_token)
+    {
+        Assembly assembly = GetAssemblyForHandle(image);
+        return RuntimeTypeHandle.ToIntPtr(assembly.Modules.Single().ModuleHandle.GetRuntimeTypeHandleFromMetadataToken((int)type_token));
```

Comment:
Would it be better to use `GetLoadedModules(getResourceModules: true)[0]` instead of `Modules.Single()` to avoid calling `IEnumerable<T>.Single()`?